### PR TITLE
White Herb doubles fix

### LINKED ISSUE #2367: Known Mechanics Bugs
Anyone is encouraged to take on any of these tasks! Once a task is completed, either comment below directly, make a reference to this issue from the pull request, or feel free to check it off yourself if you're a Collaborator. Checked items will be removed after some time.
### Battle Mechanics
- [ ] **Spread moves effect order**
  This is a serious problem with how Doubles/Triples currently handles spread moves: individually hitting each target and running all the relevant effects on the first target before moving to the next one. In-game, each target goes through one step before moving to the next step and going through each target again, much like end-of-turn effects. See also [general flow of after move effects](https://web.archive.org/web/20160721020302/http://upcarchive.playker.info:80/0/upokecenter/content/pokemon-black-version-and-pokemon-white-version-after-the-attack.html).
- [ ] **Moves called by other moves should be prevented by some "BeforeMove" checks**
  Specifically, effects that stop certain moves from executing. For example, Metronome -> Recover should be stopped by Heal Block but not full paralysis. See also [submoves](https://dex.pokemonshowdown.com/articles/submoves).

- [ ] **Berry interaction with Unnerve switching to Unnerve**
  The opposing side's Berries should be able to activate between the time the first Unnerve Pokemon switches out and before the next Unnerve Pokemon switches in.

- [ ] **Power of Alchemy/Receiver interaction with Beast Boost/Moxie/Soul-Heart**
  As the former takes over the latter when an ally faints, the new Ability immediately grants a boost if the ally fainted from the user's attack (in Soul-Heart's case, even if it fainted for other reasons).

- [ ] **Turn order**
  See #1157

##### Gen IV
- [ ] **Encore should fail if the target did not make a move during its turn**
  Due to flinch, full paralysis, etc.

- [ ] **Encore targeting in Double Battles**
  When a target is affected it should not get a chance to choose a target with the forced move; target is randomly chosen.

- [ ] **U-turn into Intimidate doesn't work when breaking Substitutes**
  Reported in https://www.smogon.com/forums/threads/bug-reports-v2-0-read-op-before-posting.3469932/page-378#post-7597157

##### Gen III
--

##### Gen II
- [ ] **Implement King's Rock properly**
  See #3343 and [details](https://web.archive.org/web/20170530082915/http://upcarchive.playker.info:80/0/upokecenter/content/pokemon-gold-version-silver-version-and-crystal-version-list-of-moves.html) (King's Rock is move flag `f` on this page)

---
### Research needed
- Speed calculation in past gens
- Hacked genders in past gens (see #2327)
- Truant in past gens (see [Gen 3 footage](https://www.smogon.com/forums/threads/gen-3-on-ps-final-fixes.3527268/page-4#post-6706179))
- Which Berries cannot activate Cheek Pouch when eaten with Bug Bite/Fling/Pluck
- Similarly, which Berries do not count for Belch's requirement of having eaten a Berry to be used (when eaten with Bug Bite/Fling/Pluck). See also the [Bug Report post](http://www.smogon.com/forums/threads/bug-reports-v2-0-read-op-before-posting.3469932/page-261#post-6732695) and the one right below it.

---
### Client Stuff
- [ ] **`-cureteam` should not remove status from active Pokemon that were not actually cured**
  For example, [Aromatherapy with an ally behind a substitute](http://replay.pokemonshowdown.com/randomdoublesbattle-331578248) (Turn 2)

### PR DESCRIPTION
**Pull request by littlefoot22**
#2367 

White Herb doesn't work with doubles.

Not sure if I have the right approach here...please comment if not,

Thanks!!

### CODE REVIEW

File: test/simulator/items/whiteherb.js [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+const common = require('./../../common');
+
+let battle;
+
+describe('White Herb', function () {
+	afterEach(function () {
+		battle.destroy();
+	});
+
+	it('it should use white herb after both intimidate', function () {
```
Comment (MacChaeger) [inline test/simulator/items/whiteherb.js]: You can leave off the extra “it” here; that’s what the function name is for.

Comment (littlefoot22) [reply test/simulator/items/whiteherb.js]: Thanks !! Removed

File: data/items.js [hunk 2]
```diff
@@ -5978,14 +5978,18 @@ let BattleItems = {
 			let activate = false;
 			/**@type {{[k: string]: number}} */
 			let boosts = {};
+			//doubles
+			if (pokemon.side.foe.pokemon.length > 1 && pokemon.side.foe.battle.abilityOrder !== 4) {
```
Comment (Zarel) [inline data/items.js]:
```suggestion
			if (pokemon.side.active.length > 1 && this.abilityOrder !== 4) {
```

**Comment by Marty-D** [comment]
There are a few problems here; this kind of fix only works before Turn 1 and not when multiple Pokemon are replacing fainted ones at the end of a turn, but more importantly White Herb itself would only work while those four lead Pokemon are on the field and won't ever activate again for anything in Doubles. `abilityOrder` keeps counting up with every Ability change or switch in.

**Comment by littlefoot22** [comment]
> There are a few problems here; this kind of fix only works before Turn 1 and not when multiple Pokemon are replacing fainted ones at the end of a turn, but more importantly White Herb itself would only work while those four lead Pokemon are on the field and won't ever activate again for anything in Doubles. `abilityOrder` keeps counting up with every Ability change or switch in.

Thank you for the feed back!! I had a feeling I didn't fully understand the variables I was using, I will do more testing and work on this bug some more

**Comment by MacChaeger** [comment]
What if we made the `Update` event only happen after all switching is finished?

**Comment by MacChaeger** [comment]
If multiple Talonflames holding Sitrus Berries switch into Stealth Rock, do the berries wait until all Pokémon have switched in?

**Comment by Marty-D** [comment]
> 
> 
> If multiple Talonflames holding Sitrus Berries switch into Stealth Rock, do the berries wait until all Pokémon have switched in?

Nope, Berries activate as soon as each one takes the hazard damage, like switch-in Abilities. That's actually why Unnerve activates before hazards, so it can stop faster Pokemon from healing in those situations.

**Comment by littlefoot22** [comment]
@Marty-D @Zarel @MacChaeger updated the PR with a new approach, running the update after the queue is empty to make sure it happens after all switch-ins happen. Is there a better way to do this? Is there an issue with this approach?

**Comment by littlefoot22** [comment]
> There are a few problems here; this kind of fix only works before Turn 1 and not when multiple Pokemon are replacing fainted ones at the end of a turn, but more importantly White Herb itself would only work while those four lead Pokemon are on the field and won't ever activate again for anything in Doubles. `abilityOrder` keeps counting up with every Ability change or switch in.

Hi @Marty-D, updated the PR with a new solution and tests based on your feedback.

**Comment by Marty-D** [comment]
This new approach also doesn't work because the queue isn't empty during a turn, for example:
```
The opposing Gyarados used Sleep Talk!
But it failed!

The opposing Incineroar used Sleep Talk!
But it failed!

Clefable used Memento!
The opposing Incineroar's Attack fell harshly!
The opposing Incineroar's Special Attack fell harshly!

Clefable fainted!
Chansey used Memento!
The opposing Incineroar's Attack fell harshly!
The opposing Incineroar's Special Attack fell harshly!

Chansey fainted!

The opposing Incineroar returned its status to normal using its White Herb!
```
White Herb should activate before Chansey moves here.

**Comment by littlefoot22** [comment]
> This new approach also doesn't work because the queue isn't empty during a turn, for example:
> 
> ```
> The opposing Gyarados used Sleep Talk!
> But it failed!
> 
> The opposing Incineroar used Sleep Talk!
> But it failed!
> 
> Clefable used Memento!
> The opposing Incineroar's Attack fell harshly!
> The opposing Incineroar's Special Attack fell harshly!
> 
> Clefable fainted!
> Chansey used Memento!
> The opposing Incineroar's Attack fell harshly!
> The opposing Incineroar's Special Attack fell harshly!
> 
> Chansey fainted!
> 
> The opposing Incineroar returned its status to normal using its White Herb!
> ```
> White Herb should activate before Chansey moves here.

got it, will do more testing and work on it, appreciate you taking a look!

**Comment by littlefoot22** [comment]
> This new approach also doesn't work because the queue isn't empty during a turn, for example:
> 
> ```
> The opposing Gyarados used Sleep Talk!
> But it failed!
> 
> The opposing Incineroar used Sleep Talk!
> But it failed!
> 
> Clefable used Memento!
> The opposing Incineroar's Attack fell harshly!
> The opposing Incineroar's Special Attack fell harshly!
> 
> Clefable fainted!
> Chansey used Memento!
> The opposing Incineroar's Attack fell harshly!
> The opposing Incineroar's Special Attack fell harshly!
> 
> Chansey fainted!
> 
> The opposing Incineroar returned its status to normal using its White Herb!
> ```
> White Herb should activate before Chansey moves here.

@Marty-D I updated the PR, would something like this work? onResidual runs after both intimidates activate for doubles and we can keep the update to run only after moves that decrees stats like tailwhip/growl/stringshot etc ?

File: test/simulator/items/whiteherb.js [hunk 3]
```diff
@@ -0,0 +1,150 @@
+'use strict';
+
+const assert = require('./../../assert');
+const common = require('./../../common');
+
+let battle;
+
+describe('White Herb 5678765567', function () {
```
Comment (Zarel) [inline test/simulator/items/whiteherb.js]: Debugging artifact, I presume?

```suggestion
describe('White Herb', function () {
```

Comment (littlefoot22) [reply test/simulator/items/whiteherb.js]: yup :D will remove

File: test/simulator/items/whiteherb.js [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+
+
+		battle.makeChoices('move bodyslam, move agility', 'move gyroball, move airslash');
+
+		const holder = battle.p2.active[0];
+
+		assert.false.holdsItem(holder);
+		assert.statStage(holder, 'atk', 0);
+	});
+	it('should use white herb after two intimidate switch in', function () {
```
Comment (Marty-D) [inline test/simulator/items/whiteherb.js]: Switching in during a turn (one at a time) is fundamentally different from switching in at the beginning of a battle or to replace fainted Pokemon (all at once). This test you're doing here should activate White Herb before the second Pokemon even switches out.

Comment (littlefoot22) [reply test/simulator/items/whiteherb.js]: ah ok makes sense, will update the PR

Comment (littlefoot22) [reply test/simulator/items/whiteherb.js]: > Switching in during a turn (one at a time) is fundamentally different from switching in at the beginning of a battle or to replace fainted Pokemon (all at once). This test you're doing here should activate White Herb before the second Pokemon even switches out.

@Marty-D 

It seems runSwitch remains in queue when Pokemon are switching in at start or after feint. Checking for this before the update leads to White Herb being called after all Pokemon switch in after start or after feint and calls it right away after a player switches during a turn. Is this correct behavior?

Comment (MacChaeger) [reply test/simulator/items/whiteherb.js]:
```suggestion
	it('should use white herb after two Intimidate switch in', function () {
```

File: data/items.js [hunk 5]
```diff
@@ -6009,9 +6009,17 @@ let BattleItems = {
 			},
 		},
 		onUpdate(pokemon) {
+			if ((this.gameType === "doubles")) {
```
Comment (urkerab) [inline data/items.js]: This doesn't look like a good idea to me, since it's too specific, but do you even need it any more?

Comment (littlefoot22) [reply data/items.js]: probably not I was going to add some tests for singles and test without the check

**Comment by MacChaeger** [comment]
I wonder if it'd work to run the `SwitchIn` event the way we run spread moves. Could we pass all the currently switching Pokemon to `runEvent` in an array?

**Comment by littlefoot22** [comment]
@MacChaeger not sure I understand what your proposing, would this be to check if more Pokemon are about to be switched in ?

**Comment by AnnikaCodes** [comment]
Old PR from someone who's no longer contributing.

**Pull request closed without merging**